engine/src/main/java/io/camunda/zeebe/engine/state/message/TransientSubscriptionCommandState.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.engine.state.message;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * This class keeps track of the last sent time of commands related to either message subscriptions
+ * or process message subscriptions.
+ */
+final class TransientSubscriptionCommandState {
+
+  private final List<CommandEntry> transientState = new ArrayList<>();
+
+  final void add(final CommandEntry commandEntryToAdd) {
+    removeEqualEntry(commandEntryToAdd);
+    synchronized (transientState) {
+      transientState.add(commandEntryToAdd);
+    }
+  }
+
+  final void updateCommandSentTime(final CommandEntry updatedCommandEntry) {
+    final var existingEntry = removeEqualEntry(updatedCommandEntry);
+    if (existingEntry != null) {
+      add(updatedCommandEntry);
+    }
+  }
+
+  final void remove(final CommandEntry templateOfCommandEntryToBeRemoved) {
+    removeEqualEntry(templateOfCommandEntryToBeRemoved);
+  }
+
+  private CommandEntry removeEqualEntry(final CommandEntry templateCommandEntry) {
+    synchronized (transientState) {
+      for (final Iterator<CommandEntry> i = transientState.iterator(); i.hasNext(); ) {
+        final var entryToCompare = i.next();
+        if (entryToCompare.equals(templateCommandEntry)) {
+          i.remove();
+          return entryToCompare;
+        }
+      }
+    }
+    return null;
+  }
+
+  final Iterable<CommandEntry> getEntriesBefore(final long deadline) {
+    final var result = new ArrayList<CommandEntry>();
+    synchronized (transientState) {
+      Collections.sort(transientState);
+      for (final CommandEntry commandEntry : transientState) {
+        if (commandEntry.getCommandSentTime() >= deadline) {
+          break;
+        }
+
+        result.add(commandEntry);
+      }
+    }
+    return result;
```

Comment:
(optional) We could write this a bit shorter using streams.

```suggestion
    return transientState.stream()
          .sorted()
          .takeWhile(commandEntry -> commandEntry.getCommandSentTime() < deadline)
          .collect(Collectors.toList());
```

Reply:
Nice one!